### PR TITLE
Fixing code supporting co_await

### PR DESCRIPTION
**Pull request by hkaiser**
- adapt specializations of coroutine_traits to latest changes
- properly propagate exceptions through co_await
- flyby: fix co_await future<future<T>>
- flyby: change return to co_return where appropriate

### CODE REVIEW

File: hpx/lcos/detail/future_await_traits.hpp [hunk 1]
```diff
@@ -74,14 +83,78 @@ namespace hpx { namespace lcos { namespace detail
...
+            return std::experimental::suspend_never{};
+        }
+
+        std::experimental::suspend_if final_suspend()
+        {
+            // This gives up the coroutine's reference count on the shared
+            // state. If this was the last reference count, the coroutine
+            // should not suspend before exiting.
+            return std::experimental::suspend_if{
+                !this->base_type::requires_delete()};
```
Comment (sithhell) [inline hpx/lcos/detail/future_await_traits.hpp]: `suspend_if` doesn't seem to be part of the TS, where is it defined?

Comment (hkaiser) [inline hpx/lcos/detail/future_await_traits.hpp]: Good point, I didn't notice they removed it. We can easily add our own here.

File: hpx/lcos/detail/future_await_traits.hpp [hunk 2]
```diff
@@ -74,14 +83,78 @@ namespace hpx { namespace lcos { namespace detail
...
+
+        void set_exception(std::exception_ptr e)
+        {
+            try {
+                std::rethrow_exception(std::move(e));
+            }
+            catch (...) {
+                this->base_type::set_exception(boost::current_exception());
+            }
+        }
```
Comment (hkaiser) [inline hpx/lcos/detail/future_await_traits.hpp]: The first is for 'conformance', the second for our internal use.

**Comment by hkaiser** [comment]
> Does this break the clang implementation?

I'd assume not (as I have not changed any use of API's), but I have not checked. Could you do that?

**Comment by sithhell** [comment]
examples/transpose/transpose_await.cpp needs to be adapted as well.

**Comment by hkaiser** [comment]
@sithhell Your comments have been addressed.

**Pull request merged**